cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java

```diff
@@ -803,6 +805,100 @@ public void suspendResumeSchedule(String namespaceId, String appId, String sched
   }
 
   /**
+   * Add a schedule to an application.
+   *
+   * @param applicationId the application id for which the schedule needs to be added
+   * @param scheduleSpec the schedule specification
+   * @throws NotFoundException when application is not found
+   * @throws SchedulerException on an exception when updating the schedule
+   */
+  public void addSchedule(ApplicationId applicationId, ScheduleSpecification scheduleSpec)
+    throws NotFoundException, SchedulerException, AlreadyExistsException {
+    ApplicationSpecification appSpec = store.getApplication(applicationId);
+    if (appSpec == null) {
+      throw new ApplicationNotFoundException(applicationId);
+    }
+
+    String scheduleName = scheduleSpec.getSchedule().getName();
+    ScheduleSpecification existingScheduleSpec = appSpec.getSchedules().get(scheduleName);
+    if (existingScheduleSpec != null) {
+      throw new AlreadyExistsException(
+        new ScheduleId(applicationId.getNamespace(), applicationId.getApplication(), scheduleName));
+    }
+
+    // TODO: Figure out if there a way to make the scheduler update and store update transactional
+    ProgramType programType = ProgramType.valueOfSchedulableType(scheduleSpec.getProgram().getProgramType());
```

Comment:
This should only succeed for workflows. If it succeeds for a mapreduce, that would create the false illusion that that is possible. But most likely it will then fail when the schedule triggers.